src/util/settings.cpp

```diff
@@ -0,0 +1,161 @@
+// Copyright (c) 2019 The Bitcoin Core developers
+// Distributed under the MIT software license, see the accompanying
+// file COPYING or http://www.opensource.org/licenses/mit-license.php.
+
+#include <util/settings.h>
+
+#include <univalue.h>
+
+namespace util {
+namespace {
+
+struct Source {
+    SettingsSpan span;
+    bool forced = false;
+    bool config_file = false;
+    bool config_file_top_level = false;
+
+    explicit Source(SettingsSpan span) : span(span) {}
+    Source& SetForced() noexcept { forced = true; return *this; }
+    Source& SetConfigFile(bool top_level) noexcept { config_file = true; config_file_top_level = top_level; return *this; }
+};
+
+template <typename Fn>
+static void MergeSettings(const Settings& settings, const std::string& section, const std::string& name, Fn&& fn)
+{
+    if (auto* value = FindKey(settings.forced_settings, name)) {
+        fn(Source(SettingsSpan(*value)).SetForced());
+    }
+    if (auto* values = FindKey(settings.command_line_options, name)) {
+        fn(Source(SettingsSpan(*values)));
+    }
+    if (!section.empty()) {
+        if (auto* map = FindKey(settings.ro_config, section)) {
+            if (auto* values = FindKey(*map, name)) {
+                fn(Source(SettingsSpan(*values)).SetConfigFile(/* top_level= */ false));
+            }
+        }
+    }
+    SettingsSpan span;
+    if (auto* map = FindKey(settings.ro_config, "")) {
```

Comment:
I believe `""` are rather overused on codebases. This invokes `std::string`'s copy constructor taking `const char*` that will try to `strlen()`, etc. Creating default-constructed `std::string{}`/ `std::string()`/`{} (in right context)` would feel more "pedantically correct" :) .

Reply:
Right, I'm used to QString too much. std::string optimizes well ("" same as {}), it's QString where it does add cost (extra function all). Sorry for bothering.